src/components/BranchMenu.tsx

```diff
@@ -0,0 +1,315 @@
+import ListItem from '@material-ui/core/ListItem';
+import ClearIcon from '@material-ui/icons/Clear';
+import { showErrorMessage } from '@jupyterlab/apputils';
+import { Git, IGitExtension } from '../tokens';
+import {
+  activeListItemClass,
+  filterClass,
+  filterClearClass,
+  filterInputClass,
+  filterWrapperClass,
+  listItemClass,
+  listItemIconClass,
+  listWrapperClass,
+  newBranchButtonClass,
+  wrapperClass
+} from '../style/BranchMenu';
+import { NewBranchDialog } from './NewBranchDialog';
+
+const CHANGES_ERR_MSG =
+  'The current branch contains files with uncommitted changes. Please commit or discard these changes before switching to or creating another branch.';
+
+/**
+ * Interface describing component properties.
+ */
+export interface IBranchMenuProps {
+  /**
+   * Git extension data model.
+   */
+  model: IGitExtension;
+
+  /**
+   * Boolean indicating whether branching is disabled.
+   */
+  branching: boolean;
+}
+
+/**
+ * Interface describing component state.
+ */
+export interface IBranchMenuState {
+  /**
+   * Menu filter.
+   */
+  filter: string;
+
+  /**
+   * Boolean indicating whether to show a dialog to create a new branch.
+   */
+  branchDialog: boolean;
+
+  /**
+   * Current branch name.
+   */
+  current: string;
+
+  /**
+   * Current list of branches.
+   */
+  branches: Git.IBranch[];
+}
+
+/**
+ * React component for rendering a branch menu.
+ */
+export class BranchMenu extends React.Component<
+  IBranchMenuProps,
+  IBranchMenuState
+> {
+  /**
+   * Returns a React component for rendering a branch menu.
+   *
+   * @param props - component properties
+   * @returns React component
+   */
+  constructor(props: IBranchMenuProps) {
+    super(props);
+
+    const repo = this.props.model.pathRepository;
+    this._addListeners();
+
+    this.state = {
+      filter: '',
+      branchDialog: false,
+      current: repo ? this.props.model.currentBranch.name : '',
+      branches: repo ? this.props.model.branches : []
+    };
+  }
+
+  /**
+   * Callback invoked when a component will no longer be mounted.
+   */
+  componentWillUnmount() {
+    this._removeListeners();
+  }
+
+  /**
+   * Renders the component.
+   *
+   * @returns fragment
+   */
+  render() {
+    return (
+      <div className={wrapperClass}>
+        <div className={filterWrapperClass}>
+          <div className={filterClass}>
+            <input
+              className={filterInputClass}
+              type="text"
+              onChange={this._onFilterChange}
+              value={this.state.filter}
+              placeholder="Filter"
+              title="Filter branch menu"
+            />
+            {this.state.filter ? (
+              <button className={filterClearClass}>
+                <ClearIcon
+                  titleAccess="Clear the current filter"
+                  fontSize="small"
+                  onClick={this._resetFilter}
+                />
+              </button>
+            ) : null}
```

Comment:
Personally, I don't see this as a simplification. While using `||` is relatively common in JavaScript-land when setting default values, using `&&` in this way is not. While `||` lends itself to saying `this OR that`, permitting a mental model other than truth tables, `&&` translates to `this AND that` which people commonly (although [mistakenly](https://www.ecma-international.org/publications/files/ECMA-ST/Ecma-262.pdf)) assume will be a `boolean`.

Anecdotally, that this erroneous mental model is common is supported by the fact that I've both seen and experienced interviews in which one is, as something of a trick question to distinguish language "aficionados", asked the result type of `'' && null`.

And while Python behaves similarly to JavaScript,

```python
>>> '' and false
''
```

Julia, e.g., goes so far as to prohibit non-boolean values in boolean contexts.

```julia
julia> "" && false
ERROR: TypeError: non-boolean (String) used in boolean context
```

In short, I personally tend to favor being explicit rather than relying on language quirks that are neither widely ubiquitous (despite being shown in the React documentation) as convention nor straightforward to immediately reason about.

Reply:
@kgryte I'm not as sure about the wider JS world, but I do know that taking advantage of `&&` short-circuiting is part of the idiom of Python (it's discussed extensively even in introductory tutorials), it's part of the idiom of React, and it's also used frequently in jlab core. Whether you change to using it here or not is a minor stylistic (and maybe performance?) niggle, but in general I don't think we have any reason to shy away from using the short-circuiting syntax/semantics.

I think you can reason about the behavior if you remember two things:

- Most (all?) values in Python/JS can participate in logical operations, but they won't be automagically converted to booleans

- The point of short-circuiting is to not evaluate the second clause of a logical operator if the answer is known after evaluating the first

> asked the result type of `'' && null`

From the above, you can work out that since `''` is falsey, the second clause will be ignored and the return value will just be `''`. 

Another tricky question would be "what's the return value of `'' || null`?". I'll try to prove my point by working it out (since I don't actually know the answer): `''` is falsey, so no short circuiting happens here, the second clause will be eval-ed, and the return will be `null`.

Reply:
and I was right! yaaay

Reply:
@telamonian I think you not knowing the answer *a priori* for `'' || null` is part of my point. ;)

But recognized that logical short-circuiting is a common Python idiom and found in React. Just not convinced that because one can, and does, that one should.

Using logical short-circuiting does **not** provide a performance benefit when compared to using a ternary operator in this situation. For reference, see sections [12.13.3](https://www.ecma-international.org/publications/files/ECMA-ST/Ecma-262.pdf) and [12.14.3](https://www.ecma-international.org/publications/files/ECMA-ST/Ecma-262.pdf).